update description (#1)
DevExpress-Examples/how-to-emulate-aspxgridview-command-button-functionality-e4664: Readme.md

# Grid View for ASP.NET Web Forms - How to emulate command button functionality
<!-- run online -->
**[[Run Online]](https://codecentral.devexpress.com/e4664/)**
<!-- run online end -->

> This example demonstrates how to add custom buttons to a templated column and configure the grid's cell edit functionality based on the edit mode. You can choose the edit mode from the combo box editor.

Reviewer: > You can choose the edit mode **from** the combo box editor.

This is not a correct preposition. You do not "pick" the edit mode "from" the combo box edit, you choose or select it among other options **in** the combo box editor. You should realize where this mistake comes from -- in Russian you can say "to choose something from [the available selection]", yet in English such wording sounds strange indeed.

You can ask yourself a correct question - where you can choose the edit mode? **In** the combo box editor.

The preposition "in" is neutral, while "from" implies a movement outside - to move from, to remove from, something **away**...

I do hope this explanation makes sense. If not, you need only ask.

```suggestion
This example demonstrates how to add custom buttons to a templated column and configure the grid's cell edit functionality based on the edit mode. You can choose the edit mode in the combo box editor.
```

![Emulate Command Buttons](commandButtons.png)

## Overview

> 1. Specify a column' [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) property and add custom **New**, **Edit**, and **Delete** buttons to the template. For **Edit** and **Delete** buttons, handle their server-side `Init` events to access the button's template container and get the container's visible index. For all buttons, handle their client-side `Click` events and call a corresponding grid's method to edit data.

Reviewer: > Specify a column' [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) property

Looks like, we've got a typo here.

> For all buttons, handle their client-side `Click` events and call **a** corresponding grid's method to edit data.

Please remember this -- "corresponding" is a **very heavy modifier** that adds definiteness to anything in 100% cases. There can't be **any** corresponding method, because if there is a direct link between two entities -- match, or correspondence, it renders the article case 100% definite.

> For **Edit** and **Delete** buttons, handle their server-side `Init` events to access **the** button's template container and get the container's visible index.

We speak of multiple buttons, therefore, "template container" is related to **any** of them, correct?

```suggestion
1. Specify a column's [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) property and add custom **New**, **Edit**, and **Delete** buttons to the template. For **Edit** and **Delete** buttons, handle their server-side `Init` events to access a button's template container and get the container's visible index. For all buttons, handle their client-side `Click` events and call the corresponding grid's method to edit data.
```

                    </td>
                    <td>
                        <dx:ASPxButton ID="btnEdit" runat="server" Text="Edit" AutoPostBack="false"
                            OnInit="btnEdit_Init" />
                    </td>
                    <td>
                        <dx:ASPxButton ID="btnDelete" runat="server" Text="Delete" AutoPostBack="false"
                            OnInit="btnDelete_Init" />
                    </td>
                </tr>
            </table>
        </DataItemTemplate>
        <!-- ... -->
    </dx:GridViewDataColumn>
    ```

    ```csharp
    protected void btnEdit_Init(object sender, EventArgs e) {
        var btn = (sender as ASPxButton);
        var nc = btn.NamingContainer as GridViewDataItemTemplateContainer;
        btn.ClientSideEvents.Click = String.Format("function() {{ grid.StartEditRow({0}); }}", nc.VisibleIndex);
    }

    protected void btnDelete_Init(object sender, EventArgs e) {
        var btn = (sender as ASPxButton);
        var nc = btn.NamingContainer as GridViewDataItemTemplateContainer;
        btn.ClientSideEvents.Click = String.Format("function() {{ grid.DeleteRow({0}); }}", nc.VisibleIndex);
    }
    ```

> 2. Specify a column's [EditItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.EditItemTemplate) property and add custom **Update** and **Cancel** buttons to the template. Handle the client-side `Click` events and call the grid's `UpdateEdit` and `CancelEdit` in the handlers.

Reviewer: > ... and call the grid's `UpdateEdit` and `CancelEdit` methods in **the** handlers.

We told  about handling events but haven't told anything about handlers themselves. Since we lack a proper introduction, I would have treated this case indefinite. Therefore, we need no definite article here.

> Handle **the** client-side `Click` events

We've got a plural form on our hands here, therefore, we can omit the definite article. Why? Because we speak of multiple `Click` events of multiple buttons, and we need not to point out which of them.

I hope this explanation makes sense.

```suggestion
2. Specify a column's [EditItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.EditItemTemplate) property and add custom **Update** and **Cancel** buttons to the template. Handle client-side `Click` events and call the grid's `UpdateEdit` and `CancelEdit` methods in handlers.
```

    ```aspx
    <dx:GridViewDataColumn Caption="Commands">
        <!-- ... -->
        <EditItemTemplate>
            <table>
                <tr>
                    <td>
                        <dx:ASPxButton ID="btnUpdate" runat="server" Text="Update" AutoPostBack="false">
                            <ClientSideEvents Click="function() { grid.UpdateEdit(); }" />
                        </dx:ASPxButton>
                    </td>
                    <td>
                        <dx:ASPxButton ID="btnCancel" runat="server" Text="Cancel" AutoPostBack="false">
                            <ClientSideEvents Click="function() { grid.CancelEdit(); }" />
                        </dx:ASPxButton>
                    </td>
                </tr>
            </table>
        </EditItemTemplate>
        <EditFormSettings Visible="false" />
    </dx:GridViewDataColumn>
    ```

> 3. For the grid's edit form mode, specify the control's [Templates.EditForm](v) property and set the [ReplacementType](https://docs.devexpress.com/AspNet/DevExpress.Web.ASPxGridViewTemplateReplacement.ReplacementType) property to `EditFormEditors`. Add custom **Update** and **Cancel** buttons to the template and handle their client-side `Click` events as described in the previous step.

Reviewer: > For the grid's edit form mode

Could you please clarify what do you mean by this phrase?

Did you want to say "to enable the edit form mode" or "to configure the edit form mode in the grid"? Something else, perhaps?

Right now, this phrase is ambiguous and lacks proper clarity.

Author: In this step, I describe how to add custom buttons to the grid's built-in edit form. So I think the following change makes sense (I also fixed the link to the API member):
```suggestion
3. To configure the grid's edit functionality in edit form mode, specify the control's [Templates.EditForm](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewTemplates.EditForm) property and set the [ReplacementType](https://docs.devexpress.com/AspNet/DevExpress.Web.ASPxGridViewTemplateReplacement.ReplacementType) property to `EditFormEditors`. Add custom **Update** and **Cancel** buttons to the template and handle their client-side `Click` events as described in the previous step.
```

Reviewer: Yes, this change makes sense indeed. Thank you!
